cpp/src/arrow/dataset/filter.cc

```diff
@@ -882,6 +963,36 @@ Result<Datum> TreeEvaluator::Evaluate(const NotExpression& expr,
   return std::move(out);
 }
 
+Result<Datum> TreeEvaluator::Evaluate(const InExpression& expr,
+                                      const RecordBatch& batch) const {
+  ARROW_ASSIGN_OR_RAISE(auto operand_values, Evaluate(*expr.operand(), batch));
+  if (IsNullDatum(operand_values)) {
+    return Datum(expr.set()->null_count() != 0);
+  }
+
+  DCHECK(operand_values.is_array());
+  Datum out;
+  compute::FunctionContext ctx{pool_};
+  RETURN_NOT_OK(arrow::compute::IsIn(&ctx, operand_values, expr.set(), &out));
+  return std::move(out);
+}
+
+Result<Datum> TreeEvaluator::Evaluate(const IsValidExpression& expr,
+                                      const RecordBatch& batch) const {
+  ARROW_ASSIGN_OR_RAISE(auto operand_values, Evaluate(*expr.operand(), batch));
+  if (IsNullDatum(operand_values)) {
+    return Datum(false);
+  }
+
+  if (operand_values.is_scalar()) {
+    return Datum(true);
+  }
+
+  DCHECK(operand_values.is_array());
+  return Datum(std::make_shared<BooleanArray>(operand_values.array()->length,
+                                              operand_values.array()->buffers[0]));
```

Comment:
This may construct an array with null data, if the operand has no nulls.

Reply:
thanks, I'll add a check for null_count == 0